ap/columnar_util.py

```diff
@@ -1132,11 +1132,32 @@ def run_setup(
         # store the call kwargs
         self.call_kwargs = call_kwargs
 
-    def __call__(self, *args, **kwargs):
+    def __call__(self, *args, call_cache=None, call_force=False, **kwargs):
         """
         Calls the wrapped function with all *args* and *kwargs*. The latter is updated with
         :py:attr:`call_kwargs` when set, but giving priority to existing *kwargs*.
+
+        Also, all calls are cached unless *call_cache* is *False*. In case caching is active and
+        this instance was called before (identified by its class and :py:attr:`name`), it is not
+        called again but *None* is returned. This check can be bypassed for this call only by
+        setting *call_force* to *True*.
         """
+        # call caching
+        cache_kwargs = {}
+        if call_cache is False:
+            # setup a new call cache when not present yet
+            if call_cache is None:
+                call_cache = defaultdict(int)
+
+            # check if the instance was called before
+            cache_key = (self.__class__, self.__name__)
+            if call_cache[cache_key] > 0 and not call_force:
+                return
+
+            # increase the count and set kwargs for the call downstream
+            call_cache[cache_key] += 1
+            cache_kwargs["call_cache"] = call_cache
```

Comment:
Ah, I forgot to add the `cache_kwargs` to the call in the first place. It's added now and the caching works! Infamous "last minute changes" ...

> Thinking about this, wouldn't it be better to implement the call cache at the class level in some way and accessing it via self.call_cache, rather than creating it here as a local and passing it around between function calls?

I was thinking about this too, but I hesitated since the cache's lifetime is limited to the call, and if added as an instance attribute, wouldn't be thread-safe (not sure if we need that though).

Reply:
Yes, I see the point about thread safety, then that makes sense of course. Adding the `cache_kwargs` could solve things, so if you didnsome testing this should be fine.